Core/Src/Scheduler/Scheduler.c

```diff
@@ -11,23 +11,21 @@
  * Initializes the priority queue and schedules tasks based on the given sensors
  * and their update frequencies. Limits the number of sensors to MAX_SENSORS.
  */
-void SchedulerInit(Scheduler* scheduler, Sensor* sensorArray[], int numSensors) {
-    if (numSensors > MAX_SENSORS) {
-        printf("Warning: Number of sensors exceeds MAX_SENSORS. Some sensors will not be scheduled.\n");
-        numSensors = MAX_SENSORS;  // Limit the number of sensors to MAX_SENSORS
-    }
-
+void SchedulerInit(Scheduler* scheduler, Updateable* updatableArray[]) {
     PQInit(&scheduler->tasks);
     scheduler->running = false;
 
-    for (int i = 0; i < numSensors; i++) {
-        Task task;
-        Sensor* sensor = sensorArray[i];
-        if (sensor == NULL) {
-            continue;
+    int count = 0;
+    for (Updateable* updateable; updateable != NULL; updateable++) {
+        count++;
+        if (count >= MAX_SENSORS) {
+            printf("Warning: Number of sensors exceeds MAX_SENSORS. Some sensors will not be scheduled.\n");
+            break;
         }
-        TaskInit(&task, sensor, sensor->updateable.hz);
-        int hz = 1000 / sensor->updateable.hz;
+
+        Task task;
+        TaskInit(&task, updateable, updateable->hz);
+        int hz = 1000 / updateable->hz;
         int initialPriority = clock() / (CLOCKS_PER_SEC / 1000) + hz;
```

Comment:
```clock()``` is a standard C library function meant for desktop systems, not meant for embedded systems, 
so if implemented it may not have the resolution needed. 
As it has been identified already, ```HAL_GetTick();``` could potentially be more suitable as a proper hardware timer.

I assume ```clock() / (CLOCKS_PER_SEC / 1000)``` tries to convert ```clock``` into miliseconds. 
```HAL_GetTick()```  get current time in milliseconds.

Suggestion:
```
uint32_t currentTime = HAL_GetTick();  // Get current time in milliseconds
uint32_t period = 1000 / updateable->hz;  // Calculate period in milliseconds
uint32_t initialExecTime = currentTime + period;  // First execution should be one period from currentTime

// initialExecTime becomes the priority - earlier times = higher priority
PQPush(&scheduler->tasks, task, initialExecTime);
```

Reply:
I think a proper solution to this may be having a flag in the main file for if code is being compiled by firmware team (onto the STM chip) or software team (for C development) and choose either clock() or HAL based off a conditional in here. And then we can use that flag in other places as this extends.

Reply:
Yeah that could work. Platform specific builds can allow firmware team and software team to develop/test on suitable timers. 

Also, I can't comment on ```SchedulerRun``` I assume because it's not part of this pull request, 
but I noticed the current calculation for execution time may not be accounting for timer overflow.

A fix could be writing a function to calculate the wrapped difference between two points with overflow in consideration, like:
```
uint32_t getTimeDifference(uint32_t start, uint32_t end) {
    if (end < start) {
        // Overflow occurred
        // Calculate wrapped difference
        return (UINT32_MAX - start) + end + 1;
    }
    return end - start;
}
```
and use it for calculating execution times, for example:
```
while (scheduler->running) {
        uint32_t currentTime = HAL_GetTick();
        
        while (!PQIsEmpty(&scheduler->tasks)) {
            if (!PQPeek(&scheduler->tasks, &currentTask)) {
                break;
            }

            // Use getTimeDifference to safely check if task is due
            uint32_t timeUntilNext = getTimeDifference(currentTime, 
                                                      currentTask.nextExecTime);
            
            if (timeUntilNext > 0) {
                break;  // Not time yet
            }

            if (PQPop(&scheduler->tasks, &currentTask)) {
                TaskExecute(&currentTask);
                
                // Schedule next execution
                currentTask.nextExecTime = currentTime + (1000 / currentTask.hz);
                PQPush(&scheduler->tasks, currentTask, currentTask.nextExecTime);
            }
        }
        
        scheduler->lastTickCount = currentTime;
```